### ISSUE TITLE
Missing labels in batches when using copy_paste and mosaic

### ISSUE THREAD

**Issue by uuu686** [ultralytics/ultralytics#18660]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

hello,
when I use my custom data to train yolo11 segment task, I find something wrong with the train_batch.jpg. Just like
![111](https://github.com/user-attachments/assets/7f977639-b791-40e4-bf96-6b33b574d45d)
The chips label is missing.
I have checked my dataset label and there is no problem, and why is that?
Now because the data split results are inconsistent, I can not reproduce for the time being.
Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @uuu686, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, it would be very helpful if you could provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug the issue. Specifically, if there are inconsistencies in the data splitting results, please try to replicate the setup where this occurs and share any relevant code, configuration files, or dataset information.

If this is a custom training ❓ Question, please provide as much information as possible, including screenshots (like the one attached), dataset examples, as well as training logs. Additionally, ensure you're adhering to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

### Clarifications and Recommendations
- Ensure your environment uses the latest `ultralytics` package and dependencies by upgrading:
```bash
pip install -U ultralytics
```
- This ensures that your issue has not already been resolved in the latest release.

### Join the Community
We encourage you to leverage the Ultralytics community for further assistance and lively discussions. Hop into:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time support
- [Discourse](https://community.ultralytics.com) for detailed discussions
- [Subreddit](https://reddit.com/r/Ultralytics) for engaging community threads

## Run Environments
YOLO may also be tested seamlessly in verified environments that come pre-configured, such as:
- Free GPU-powered **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Cloud Services: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 🌥️ or [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Pre-integrated Docker Images: <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Lastly, rest assured that this is an automated response 😊, and an Ultralytics engineer will review your issue and assist you shortly!

**Comment by Y-T-G** [comment]
What version of Ultralytics is this?

**Comment by uuu686** [comment]
I did not use the Ultralytics pip package, I just clone the default repository to do this.

**Comment by Y-T-G** [comment]
It still has a version number. Did you install it with `pip install .`?

**Comment by uuu686** [comment]
No, I didn't use that command. I mean I use "git clone xxx" to download the code.

**Comment by Y-T-G** [comment]
After that, you have to install it by running `pip install .` (dot at end) inside the cloned directory. Ultralytics is a package, not a script. You still have to install your cloned version.

**Comment by uuu686** [comment]
It's ultralytics=8.3.21. I'll see if that happens again. Thanks

**Comment by glenn-jocher** [comment]
Thank you for confirming the version. If the issue persists, please provide more details, including a minimal reproducible example, and we’ll assist further.

**Comment by uuu686** [comment]
Hello, this problem has come up again. My train.py is as follows:
`
  model.train(
      data=r'data/yamls/drink.yaml',
      epochs=4000, 
      patience=50,
      batch=4, 
      imgsz=640,  
      # imgsz=1280,  
      save=True,  
      save_period=-1,  
      cache=False,  
      # device='cuda:0',
      workers=8,  
      project='runs/train/drink/',  
      name='exp',  
      exist_ok=False,  
      pretrained=True, 
      # optimizer='SGD', 
      optimizer='auto',
      verbose=True, 
      seed=0,  
      deterministic=True,
      single_cls=False,  
      rect=False,  
      cos_lr=False,  
      close_mosaic=10,  
      resume=False,  
      amp=True,  
      fraction=1.0, 
      profile=False,  
      freeze=None,  
      # 
      overlap_mask=True,  
      mask_ratio=4,  
      # 
      dropout=0.0,  
      lr0=0.01,  
      lrf=0.01,  
      momentum=0.937,
      weight_decay=0.0005,  
      warmup_epochs=3.0,  
      warmup_momentum=0.8, 
      warmup_bias_lr=0.1,  
      box=7.5,  
      cls=0.5, 
      dfl=1.5,  
      pose=12.0,  
      kobj=1.0,  
      label_smoothing=0.0,  
      nbs=64,  
      hsv_h=0.015,  
      hsv_s=0.7,  
      hsv_v=0.4, 
      degrees=180,  
      translate=0.1,  
      scale=0.5,  
      shear=0.0, 
      perspective=0.0,  
      flipud=0.5, 
      fliplr=0.5,  
      mosaic=0.5,  
      mixup=0.0,  
      copy_paste=1.0, 
  )
`

the train_batch image is :

![Image](https://github.com/user-attachments/assets/cb55cc1c-9e0e-4cfa-82ac-51f83abaaa4f)


if I remove the mosaic, it will be ok:

![Image](https://github.com/user-attachments/assets/ae428685-0fa3-4215-86cb-bc37795fe334)

**Comment by uuu686** [comment]
this dataset can reproduce the problem
[11.zip](https://github.com/user-attachments/files/19203562/11.zip)

Thanks.

**Comment by glenn-jocher** [comment]
@uuu686 thanks for sharing the dataset. The missing labels during mosaic augmentation are likely due to aggressive translation/scale parameters in your hyperparameters (`degrees=180`, `translate=0.1`, `scale=0.5`). These can sometimes crop labels out of frame when combined with mosaic. We recommend:

1. Reduce augmentation strength:
```python
degrees=10,  # instead of 180
translate=0.05,  # instead of 0.1
scale=0.2,  # instead of 0.5
```

2. Verify your labels are centered and not near image edges where mosaic might crop them. For reference, see the Train Settings documentation on augmentation parameters.

Let us know if these adjustments resolve the issue.

**Comment by uuu686** [comment]
hello,
I tried the above parameter Settings, but the problem still exists.

![Image](https://github.com/user-attachments/assets/86dd8593-aa45-42ab-a49b-03b7cbc1df1a)

**Comment by glenn-jocher** [comment]
Thanks for following up. This appears to be caused by labels getting cropped during mosaic augmentation when using extreme geometric transformations. We recommend:

1. Verify your labels are well-centered in source images (not near edges)
2. Try reducing `copy_paste` from 1.0 to 0.5
3. Temporarily set `close_mosaic=0` to test without mosaic disable logic

If the issue persists, please share your exact current augmentation parameters and we'll help investigate further. The Mosaic augmentation implementation details might be relevant here.

**Comment by gjtjx** [comment]
This bug arises from unhandled keypoints coordinates during geometric transformations in data/augment.py

**Comment by uuu686** [comment]
> This bug arises from unhandled keypoints coordinates during geometric transformations in data/augment.py

  Excuse me. How should I fix this bug? @gjtjx

**Comment by gjtjx** [comment]
Maybe you can hava look at https://github.com/gjtjx/ultralytics/blob/main/ultralytics/data/augment.py

**Comment by uuu686** [comment]
> Maybe you can hava look at https://github.com/gjtjx/ultralytics/blob/main/ultralytics/data/augment.py

@gjtjx  I'll have a look. Thank you very much

**Comment by uuu686** [comment]
> Maybe you can hava look at https://github.com/gjtjx/ultralytics/blob/main/ultralytics/data/augment.py

@gjtjx  I replace the augment.py with your code, but the problem remains the same.

**Comment by gjtjx** [comment]
@uuu686  your augmented is hsv_h=0.015,   hsv_s=0.7,   hsv_v=0.4,  degrees=180,   translate=0.1,   scale=0.5,   shear=0.0,  perspective=0.0,   flipud=0.5,  fliplr=0.5,   mosaic=0.5,   mixup=0.0,   copy_paste=1.0？ I have check this augment on my dataset，it is run correct。can you give me ten train images，I will test it

**Comment by uuu686** [comment]
@gjtjx I constructed a simple dataset. 

[11.zip](https://github.com/user-attachments/files/20856701/11.zip)

This problem occurred during the test. 

![Image](https://github.com/user-attachments/assets/ac0f9f22-cb55-4b18-8727-7677fe5de714)

Could you please take a look, the parameter configuration is just as you said. Thanks

**Comment by gjtjx** [comment]
can you send your data.yaml?
@uuu686

**Comment by uuu686** [comment]
@gjtjx 
just like this:

`names:

  0: Wahaha_AD
  1: Oolong
  2: Coconut_water
  3: Coca_cola
  4: Pepsi_cola
  5: Alienergy
  6: Energetic_forest_orange
  7: Energetic_forest_grape
  8: Energetic_forest_peach
  9: Energetic_forest_litchi
  10: Oreo
  11: Chips_can
  12: Chips_bag
  13: scream
  14: chip_can_ori
  15: melon_seeds
  16: panda
  17: orange`

**Comment by gjtjx** [comment]
@uuu686 maybe like this 
flip_idx:
- 1
- 0
- 3
- 2
kpt_shape:
- 4
- 3
names:
- Wahaha_AD
- Oolong
- Coconut_water
- Coca_cola
- Pepsi_cola
- Alienergy
- Energetic_forest_orange
- Energetic_forest_grape
- Energetic_forest_peach
- Energetic_forest_litchi
- Oreo
- Chips_can
- Chips_bag
- scream
- chip_can_ori
- melon_seeds
- panda
- orange
nc: 18
？

**Comment by uuu686** [comment]
@gjtjx  My task is instance segmentation, not for pose. So it may don't have 'flip_idx' and  'kpt_shape'.

**Comment by gjtjx** [comment]
maybe this problem is same, mosaic is not finish  instance segmentation transform

**Comment by uuu686** [comment]
@gjtjx  Should I debug the augment.py for instance segmentation, just like your code for pose?

**Comment by Y-T-G** [comment]
@uuu686 Thanks for sharing the dataset. I can reproduce the issue with your dataset. I will debug it.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #21262, authored by Y-T-G with contributions from Laughing-q.

Key updates from this PR include enhancements to the CopyPaste augmentation—specifically, it now ensures that non-mosaic images are copied before applying CopyPaste. This prevents any direct modification of the original image, protecting your original image data and improving training stability. These changes are designed to make data augmentation more reliable and minimize unexpected side effects during training.

To test if the changes resolve your issue, you can install the latest version directly from the main branch using:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, you can wait for the next official PyPI release if you prefer to use stable releases.

Feedback on whether this update resolves your issue would be greatly appreciated. Thank you for reporting this—if you encounter any further problems, please let us know!
